contracts/orderbook/OrderBook.sol

```diff
@@ -0,0 +1,325 @@
+			keeper.call{value:gasUsed}("");
+			uint depositAmount = address(this).balance;
+			WBNB(BNBAddress).deposit{value:address(this).balance}();
+			_safeTransfer(BNBAddress,trader,IERC(BNBAddress).balanceOf(address(this)),"");
+		}
+		
+		
+        success = true;
+    }
+	function getSwapAddress() public view returns(address){
+		return StateI(bZxRouterAddress).swapsImpl();
+	}
+    function currentSwapRate(address start, address end) public view returns(uint executionPrice){
+        (executionPrice,)=IPriceFeeds(getFeed()).queryRate(start,end);
+    }
+    function getFeed() public view returns (address){
+        return StateI(bZxRouterAddress).priceFeeds();
+    }
+    function getRouter() public view returns (address) {
+        return bZxRouterAddress;
+    }
+	function marketOpen(bytes32 loanID, uint256 leverage, uint256 loanTokenAmount, uint256 collateralTokenAmount, address iToken, address base, bytes memory loanData) public{
+		executeMarketOpen(msg.sender,loanID,leverage,loanTokenAmount,collateralTokenAmount,iToken,base,loanData);
+	}
+	function marketClose(bytes32 loanID, uint amount, bool iscollateral,address loanTokenAddress, address collateralAddress,bytes memory arbData) public{
+		executeMarketClose(msg.sender,loanID,amount,iscollateral,loanTokenAddress,collateralAddress,arbData);
+	}
+    function placeOrder(IWalletFactory.OpenOrder memory Order) public{
+		require(Order.loanTokenAmount == 0 || Order.collateralTokenAmount == 0); 
+        require(currentSwapRate(Order.loanTokenAddress,Order.base) > 0);
+		require(Order.orderType <= 2);
+		require(Order.orderType > 0 ? collateralTokenMatch(Order) && loanTokenMatch(Order) : true);
+		require(Order.orderType == 0 ? Order.loanID.length == 0 || isActiveLoan(Order.loanID) : isActiveLoan(Order.loanID));
+		require(Order.loanID.length != 0 ? getTrades.inVals(ActiveTrades[msg.sender],Order.loanID) : true);
+        HistoricalOrderIDs[msg.sender]++;
+		mainOBID++;
+        Order.orderID = HistoricalOrderIDs[msg.sender];
+        Order.trader = msg.sender;
+		Order.isActive = true;
+		Order.loanData = "";
+        HistoricalOrders[msg.sender][HistoricalOrderIDs[msg.sender]] = Order;
+		AllOrders[mainOBID].trader = msg.sender;
+		AllOrders[mainOBID].orderID = Order.orderID;
+        require(sortOrderInfo.addOrderNum(HistOrders[msg.sender],HistoricalOrderIDs[msg.sender]));
+		require(sortOrderInfo.addOrderNum(AllOrderIDs,mainOBID));
+		matchingID[msg.sender][HistoricalOrderIDs[msg.sender]] = mainOBID;
+		if(getActiveTraders.inVals(activeTraders,msg.sender) == false){
+			getActiveTraders.addTrader(activeTraders,msg.sender);
+		}
+        emit OrderPlaced(msg.sender,Order.orderType,Order.price,HistoricalOrderIDs[msg.sender],Order.base,Order.loanTokenAddress);            
+    }
+    function amendOrder(IWalletFactory.OpenOrder memory Order,uint orderID) public{
+		require(Order.loanTokenAmount == 0 || Order.collateralTokenAmount == 0); 

```

Comment:
would love to se some error codes

Reply:
or combine some of them toghether

Reply:
added comments for errors to provide a description of what happened.